### PR TITLE
[CR] Add CAMOUFLAGE monster flag

### PR DESCRIPTION
**Pull request by Venera3**
#### Summary

Content "Add camouflaged monsters"

#### Purpose of change

Apart from `NIGHT_INVISIBILITY` or abusing `HIDE` terrain there's no way to hide monsters at range, which contributes to the hegemony of long-range guns. This adds an alternative to turning late-game enemies bulletproof while still impacting ranged combat.

#### Describe the solution

Add the `CAMOUFLAGE` monster flag, which prevents detecting the monster farther than ```(current Perception) + (base Perception if you have the Spotting proficiency)``` tiles for players/NPCs or `vision_night` range for monsters.

TODO:

- [x] Suppress printing the target invisible monster's name when it gets attacked by another mob but you can't actually see it
- [x] Add a monster using it - Unseen Hunters (imaginative, I know) that will be part of the nether groups and might rarely spawn in the subways (I plan on doing a proper mob spawn rework after this one for them), still some tweaking necessary
- ~~Maybe add the same functionality to `AQUATIC` critters, since seeing a whole pond's fish population at range is pretty weird~~ That will deserve to be its own change, and I do have some ideas in that space too.

#### Describe alternatives you've considered

Using a static value to define the range, but I wanted to make Per matter for once.
Doing something properly fancy and adding a JSON-definable range modifier, but I wanted to quit while I was ahead.
Have spotting/failing to spot a camouflaged critter increase the spotting proficiency, but that sounds a bit too complex for my tastes, and suddenly gaining prof experience would give them away too early.

#### Testing

Monsters with the flag are just invisible enough:
- They are visible at the intended range, 
- They disappear outside of range,
- `WARM` mobs show up in IR vision,
- Those with `NIGHT_INVISIBILITY` are invisible up to their camouflage range in good lighting, become invisible at night,
- The proficiency scaling works as intended, the detection range will ignore perception debuffs,
- Monsters detect them at their lower vision range,

Footsteps, hit messages show up according to visibility ("Something" hits stuff if you see the attacker's position but not the critter itself).

- [x] Check for any weirdness if `NIGHT_INVISIBILITY` is also present - night invisible monsters stay invisible to range 1 at night, to camouflage range in daylight
- [x] Test if footsteps appear while the monster is invisible
- [x] Mob testing when the mob actually exists

#### Additional context

Unseen hunters spawn in the `NETHER`, `NETHER_FATIGUE_FIELD` and `NETHER_PORTAL` existing monstergroups (meaning mostly around the portal map extras or from late-stage Flaming Eyeitis) after a month-long grace period. The new and pretty empty `SUBWAY` group will be fleshed out in the followup PR, for now you will see one out of every twenty subway tiles. Since you can easily generate the underlying subways during surface exploration I decided to restrict them from spawning in the first week only (to give lab escapees a bit of a chance if they want to bounce over the subways).

Balance-wise I am very happy with how they came out: they can be scared off with a good weapon (they haven't killed a single mid-game character in testing), but they regenerate in the dark and chasing them down can be a challenge. However, if you know one is lurking around you can negate both their invisibility and their regeneration with some flares and IR goggles, rewarding gearing up properly for your subway jaunts. To negate the usual fire cheese they are made of 100% fire retardant nether stuff and get even angrier if you try that old trick. Their ranged pull attack can also yoink you off a draisine or a foldable unicycle you thought would make lab looting easier, so that's an extra bit of evil sprinkled on top.

### CODE REVIEW

**Comment by actual-nh** [comment]
> `vision_night` range for monsters.

You might want to use the lower of daytime and nighttime vision range for monsters - some do have the latter higher than the former, and it'd be a tad bit odd for them to see critters with this _better_ in the daytime.

File: doc/JSON_FLAGS.md [hunk 1]
```diff
@@ -920,6 +920,7 @@ Other monster flags.
 - ```BORES``` Tunnels through just about anything (15x bash multiplier: dark wyrms' bash skill 12->180)
 - ```CAN_DIG``` Can dig _and_ walk.
 - ```CAN_OPEN_DOORS``` Can open doors on its path.
+- ```CAMOUFLAGE``` Stays invisible up to (current Perception + base Perception if the character has the Spotting proficiency) tiles away, even in broad daylight. Monsters see it from `vision_night` range.
```
Comment (actual-nh) [inline doc/JSON_FLAGS.md:923]:
```suggestion
- ```CAMOUFLAGE``` Stays invisible up to (current Perception, + base Perception if the character has the Spotting proficiency) tiles away, even in broad daylight. Monsters see it from the lower of `vision_day` and `vision_night` ranges.
```

File: src/monster.cpp [hunk 2]
```diff
@@ -1292,6 +1292,11 @@ int monster::hp_percentage() const
     return get_hp( bodypart_id( "torso" ) ) * 100 / get_hp_max();
 }
 
+int monster::get_eff_per() const
+{
+    return type->vision_night;
```
Comment (actual-nh) [inline src/monster.cpp:1297]:
```suggestion
    return std::min( type->vision_night, type->vision_day );
```

**Review by actual-nh — COMMENTED** [review]
Very nice! You might want to use the lower of daytime and nighttime vision range for monsters - some do have the latter higher than the former, and it'd be a tad bit odd for them to see critters with this better in the daytime.

**Comment by souricelle** [comment]
I wonder if this couldn't just replace night_invisibility since it's a lot more grounded. Can/does it work better at night?

**Comment by Venera3** [comment]
> 
> 
> I wonder if this couldn't just replace night_invisibility since it's a lot more grounded. Can/does it work better at night?

They work perfectly fine in tandem, but hooking up some light-based scaling to it would require fancier C++ than what I can provide.

**Comment by actual-nh** [comment]
This may need a bit of adjusting (in terms of line placement in the C++ code) due to #48353... sorry!

**Comment by actual-nh** [comment]
> Maybe add the same functionality to AQUATIC critters, since seeing a whole pond's fish population at range is pretty weird

That makes sense, at least if they're in a water spot with more than one z-level. Some things could have CAMOUFLAGE_AQUATIC, as in when they were in any depth of water (probably including swamp) they would have CAMOUFLAGE. I'm thinking of things like alligators that submerge with just their nostrils sticking out.

**Comment by Venera3** [comment]
> 
> 
> > Maybe add the same functionality to AQUATIC critters, since seeing a whole pond's fish population at range is pretty weird
> 
> That makes sense, at least if they're in a water spot with more than one z-level. Some things could have CAMOUFLAGE_AQUATIC, as in when they were in any depth of water (probably including swamp) they would have CAMOUFLAGE. I'm thinking of things like alligators that submerge with just their nostrils sticking out.

That's a good idea! I wanted to take a crack at adding some more water critters (crayfish just being out and about in the swamps is...weird) so spice up late-game boating one of these days anyway. That would also potentially let me do away with the "X dives into the deep water!" spam around water.

**Comment by Venera3** [comment]
Unseen hunters are now at least half-ready. I'm a bit torn on where in the difficulty curve I want to add them, honestly, but I think I'll aim for "bloodies a mid-game character thoroughly". Their low-ish vision range allows for you to just run away screaming, but if you start messing with the invisible horror without being prepared (or being a lab escapee, but a challenge is a challenge) you will kinda ruin that chance. To that effect I might add a slowing component to their venom, but it will be a while before I do the proper combat testing to see if that's necessary. I'm also considering giving them a shadow-emitting spell to horror it up properly, but that's only on the drawing board.

**Comment by actual-nh** [comment]
> I'm also considering giving them a shadow-emitting spell to horror it up properly, but that's only on the drawing board.

A nice thought, but one that will require having actual specified reductions (or decreases to no more than a specific maximum) to lighting from fields - currently, you can only specify the amount of light... meaning that if you specify something other than complete darkness, you can actually _increase_ the amount of light! See #46910 for a bit more on this.

**Comment by actual-nh** [comment]
> > > Maybe add the same functionality to AQUATIC critters, since seeing a whole pond's fish population at range is pretty weird
> > 
> > That makes sense, at least if they're in a water spot with more than one z-level. Some things could have CAMOUFLAGE_AQUATIC, as in when they were in any depth of water (probably including swamp) they would have CAMOUFLAGE. I'm thinking of things like alligators that submerge with just their nostrils sticking out.
> 
> That's a good idea!

Thanks! Adjusting the default visibility of AQUATIC things could be a rather good idea for anyone using Dinoclysm - there are quite a few dinos that eat both fish and other things. (Admittedly, there is a dinosaur with AQUATIC - the adult Mosasaurus - that probably should still be visible underwater, given that it's the size of a house. AQUATIC_VISIBLE or something like that? OTOH, one could also argue that unlike many dinos it doesn't have a large back "sail" sticking out making it more visible.)

> I wanted to take a crack at adding some more water critters (crayfish just being out and about in the swamps is...weird) to spice up late-game boating one of these days anyway. That would also potentially let me do away with the "X dives into the deep water!" spam around water.

Sounds good.

**Pull request merged**